test/e2e/daemon_restart.go

```diff
@@ -0,0 +1,274 @@
+	}
+}
+
+func (r *restartDaemonConfig) String() string {
+	return fmt.Sprintf("Daemon %v on node %v", r.daemonName, r.nodeName)
+}
+
+// waitUp polls healthz of the daemon till it returns "ok" or the polling hits the pollTimeout
+func (r *restartDaemonConfig) waitUp() {
+	Logf("Checking if %v is up by polling for a 200 on its /healthz endpoint", r)
+	healthzCheck := fmt.Sprintf(
+		"curl -s -o /dev/null -I -w \"%%{http_code}\" http://localhost:%v/healthz", r.healthzPort)
+
+	err := wait.Poll(r.pollInterval, r.pollTimeout, func() (bool, error) {
+		stdout, stderr, code, err := nodeExec(r.nodeName, healthzCheck)
+		expectNoError(err)
+		if code == 0 {
+			httpCode, err := strconv.Atoi(stdout)
+			if err != nil {
+				Logf("Unable to parse healthz http return code: %v", err)
+			} else if httpCode == 200 {
+				return true, nil
+			}
+		}
+		Logf("node %v exec command, '%v' failed with exitcode %v: \n\tstdout: %v\n\tstderr: %v",
+			r.nodeName, healthzCheck, code, stdout, stderr)
+		return false, nil
+	})
+	expectNoError(err, "%v did not respond with a 200 via %v within %v", r, healthzCheck, r.pollTimeout)
+}
+
+// kill sends a SIGTERM to the daemon
+func (r *restartDaemonConfig) kill() {
+	Logf("Killing %v", r)
+	nodeExec(r.nodeName, fmt.Sprintf("pgrep %v | xargs -I {} sudo kill {}", r.daemonName))
+}
+
+// Restart checks if the daemon is up, kills it, and waits till it comes back up
+func (r *restartDaemonConfig) restart() {
+	r.waitUp()
+	r.kill()
+	r.waitUp()
+}
+
+// replacePods stores the given pods in the given store.
+func replacePods(pods []*api.Pod, store cache.Store) {
+	found := make([]interface{}, 0, len(pods))
+	for i := range pods {
+		found = append(found, pods[i])
+	}
+	expectNoError(store.Replace(found))
+}
+
+// getContainerRestarts returns the count of container restarts across all pods matching the given labelSelector,
+// and a list of nodenames across which these containers restarted.
+func getContainerRestarts(c *client.Client, ns string, labelSelector labels.Selector) (int, []string) {
+	pods, err := c.Pods(ns).List(labelSelector, fields.Everything())
+	expectNoError(err)
+	failedContainers := 0
+	containerRestartNodes := util.NewStringSet()
+	for _, p := range pods.Items {
+		for _, v := range FailedContainers(&p) {
+			failedContainers = failedContainers + v.restarts
+			containerRestartNodes.Insert(p.Spec.NodeName)
+		}
+	}
+	return failedContainers, containerRestartNodes.List()
+}
+
+var _ = Describe("DaemonRestart", func() {
+
+	framework := Framework{BaseName: "daemonrestart"}
+	rcName := "daemonrestart" + strconv.Itoa(numPods) + "-" + string(util.NewUUID())
+	labelSelector := labels.Set(map[string]string{"name": rcName}).AsSelector()
+	existingPods := cache.NewStore(cache.MetaNamespaceKeyFunc)
+	var ns string
+	var config RCConfig
+	var controller *controllerFramework.Controller
+	var newPods cache.Store
+	var stopCh chan struct{}
+
+	BeforeEach(func() {
+
+		// These tests require SSH
+		if !providerIs("gce") {
```

Comment:
Can I enable it on gke in a follow up pr? don't have a cluster now and would like to see if it flakes on gce to begin with anyway. Added a TODO.